### PR TITLE
feat: Remove deprecated API use

### LINKED ISSUE #4: Broken since latest ffmpeg upgrade
Ffmpeg removed certain APIs (such as register_all) when they upgraded to support ffmpeg 5, 0.1.8 should account for those changes

### PR DESCRIPTION
**Pull request by Soremwar**
Closes #4

### CODE REVIEW

File: src/encode.rs [hunk 1]
```diff
@@ -359,8 +360,7 @@ impl Encoder {
   /// 
   /// Raw ffmpeg encoder belonging to given stream.
   fn encoder(writer_stream: &StreamMut) -> Result<AvEncoder> {
-    writer_stream
-      .codec()
+    Context::from_parameters(writer_stream.parameters())?
```
Comment (Soremwar) [inline src/encode.rs:363]: As mentioned in https://github.com/zmwangx/rust-ffmpeg/blob/feae6b21b8c9f76d399260f812c0deb620bf02ad/CHANGELOG.md

File: src/init.rs [hunk 2]
```diff
@@ -1,28 +1,14 @@
 extern crate ffmpeg_next as ffmpeg;
 
 use ffmpeg::{
-  format::{
-    version,
-    register_all,
```
Comment (Soremwar) [inline src/init.rs:6]: Removed in latest ffmpeg

**Comment by Soremwar** [comment]
You probably have to enable the workflow before it can run, but should take care of testing that everything works fine (or at least builds) under latest ffmpeg_next

File: src/init.rs [hunk 3]
```diff
@@ -1,28 +1,14 @@
...
-const AVFORMAT_MAX_VERSION_REQUIRES_REGISTER_ALL: u32 = 3803492;
-
 /// Initialize global ffmpeg settings. This also intializes the
 /// logging capability and redirect it to `tracing`.
-/// 
-/// In older versions, this will invoke `register_all`.
 pub fn init() {
-  // Older versions of ffmpeg require this.
-  if version() < AVFORMAT_MAX_VERSION_REQUIRES_REGISTER_ALL {
-    register_all();
```
Comment (gerwin3) [inline src/init.rs:23]: Is it possible to still have this check when compiling for ffmpeg4 so we retain compatibility with older ffmpeg versions?

Comment (Soremwar) [reply src/init.rs:23]: In 5.0 it was removed, but kept around under a compile flag for backwards compatibility

A compile flag could similarly be added here, but that would require all <5.0 or 5.0 users to specify it. What you think?

Comment (Soremwar) [reply src/init.rs:23]: Scratch that, apparently ffmpeg calls `register_all` automatically when it detects that ffmpeg is <5.0 on init. So this is safe to remove

https://github.com/zmwangx/rust-ffmpeg/blob/8bb9f7c7ee649a024f701aebddc79878fae4a4b5/src/lib.rs#L109-L118

Comment (Soremwar) [reply src/init.rs:23]: Added ffmpeg 3.4 to test pipeline just to be sure though

File: Cargo.toml [hunk 4]
```diff
@@ -12,6 +12,6 @@ readme = "README.md"
 
 [dependencies]
 tracing = "0.1"
-ffmpeg-next = "5.1"
+ffmpeg-next = { version = "5.1", features = [ "format" ] }
```
Comment (Soremwar) [inline Cargo.toml:15]: If other features are enabled, `init` will have additional effects

**Comment by gerwin3** [comment]
It should be possible to do this conditionally like this:

https://github.com/zmwangx/rust-ffmpeg/blob/22ad8b959879efa028c33f71cc286b519eae8066/src/codec/decoder/video.rs#L3

Those features are defined by ffmpeg-next-sys in `build.rs`, and ffmpeg-next seems to pull them from environment variables (??):

https://github.com/zmwangx/rust-ffmpeg/blob/master/build.rs

Not sure where those env vars come from...

**Comment by Soremwar** [comment]
As mentioned above, we don't really need to do that. Running `ffmpeg::init` takes care of checking whether ffmpeg will need `register_all` or not

> It should be possible to do this conditionally like this:
Not possible, since `register_all` won't be in scope to perform the feature check. I asked everywhere and came to the conclussion that it's a Rust limitation

File: Cargo.toml (hunk 4 above)
Comment (Soremwar) [reply Cargo.toml:15]: https://github.com/Soremwar/video-rs/actions/runs/3743344474/jobs/6355402414

**Comment by gerwin3** [comment]
made some last changes (also did include some of the ffmpeg-next features that we need, not sure how it compiles without) and merged this. Let me know if this works for you!

**Comment by gerwin3** [comment]
Currently the encoding example does not run and logging fails for me :thinking:. Will bump version after I fix that when I have time.

**Pull request merged**